tests/robustness/watch.go

```diff
@@ -85,61 +85,97 @@ func validateWatchResponses(t *testing.T, responses [][]watchResponse, expectPro
+	var gotProgressNotify = false
+	var lastHeadRevision int64 = 1
+	for _, resp := range responses {
 		if resp.IsProgressNotify() && resp.Header.Revision == lastHeadRevision {
 			gotProgressNotify = true
+			break
 		}
-		if resp.Header.Revision == lastHeadRevision && len(resp.Events) != 0 {
-			t.Errorf("Got two non-empty responses about same revision")
-		}
-		if len(resp.Events) > 0 && resp.Events[0].Kv.ModRevision <= lastEventModRevision {
-			t.Errorf("Events with same revision should not be split between multiple responses, lastResponseModRevision: %d, firstEventModRevision: %d", lastEventModRevision, resp.Events[0].Kv.ModRevision)
-		}
+		lastHeadRevision = resp.Header.Revision
+	}
+	if gotProgressNotify != expectProgressNotify {
+		t.Errorf("Expected at least one progress notify: %v, got: %v", expectProgressNotify, gotProgressNotify)
+	}
+}
 
+func validateBookmarks(t *testing.T, responses []watchResponse) {
+	var lastProgressNotifyRevision int64 = 0
+	for _, resp := range responses {
 		for _, event := range resp.Events {
-			rk := revisionKey{key: string(event.Kv.Key), revision: event.Kv.ModRevision}
-			if _, found := uniqueOperations[rk]; found {
-				t.Errorf("BROKE: Within the:wq same revision key can only be modified once. Suspecting duplicate watch event. key: %q, revision: %d", rk.key, rk.revision)
-			}
-			uniqueOperations[rk] = struct{}{}
 			if event.Kv.ModRevision <= lastProgressNotifyRevision {
-				t.Errorf("BROKE: etcd will not send progress notification to a watcher until it has synced all events. So a watcher will never receive an event with a lower `ModRevision` than the last progressNotification's revision, eventRevision: %d, progressNotifyRevision: %d", event.Kv.ModRevision, lastProgressNotifyRevision)
+				t.Errorf("BROKE: Bookmarks - etcd progress notification can be used as bookmarks; they guarantee that events with revision below bookmark have been already delivered, eventRevision: %d, progressNotifyRevision: %d", event.Kv.ModRevision, lastProgressNotifyRevision)
 			}
-			if event.Kv.ModRevision != lastEventModRevision && event.Kv.ModRevision != lastEventModRevision+1 {
-				t.Errorf("Event mod revision should stay the same or increment by one, last: %d, current: %d", lastEventModRevision, event.Kv.ModRevision)
+		}
+		if resp.IsProgressNotify() {
+			lastProgressNotifyRevision = resp.Header.Revision
+		}
+	}
+}
+
+func validateOrdered(t *testing.T, responses []watchResponse) {
+	var lastEventRevision int64 = 1
+	for _, resp := range responses {
+		for _, event := range resp.Events {
+			if event.Kv.ModRevision < lastEventRevision {
+				t.Errorf("BROKE: Ordered - events are ordered by revision; an event will never appear on a watch if it precedes an event in time that has already been posted, lastRevision: %d, currentRevision: %d", lastEventRevision, event.Kv.ModRevision)
 			}
-			lastEventModRevision = event.Kv.ModRevision
+			lastEventRevision = event.Kv.ModRevision
 		}
+	}
+}
```

Comment:
`validateOrdered` seems not needed at all, it's already covered by `validateReliable`.  If we ensure each `event.Kv.ModRevision` is either `lastEventRevision` or `lastEventRevision + 1`, then it means that the events must be ordered by the ModRevision.

We can consider to rename `validateReliable` to `validateReliableAndOrdered`. WDYT?

Reply:
Should be good for now, however in future I would like to separate ordered. Ordered should mean that watch events are ordered in reference to their execution on server.